### DISCUSSION TITLE
Access current active route object

### DISCUSSION THREAD

**Discussion by mammadataei** [TanStack/router#159] [Q&A] [answered]
I need to watch for route changes and access the currently active route object. I've found the following to watch for location updates:

```ts
const router = useRouter()

useEffect(() => {
  console.log(router.state.location)
}, [router.state.location])
```

I'm not sure if it's the right way to do so, and besides, I need the route object, so I should search the routes object which is quite expensive. Is there any better solution?

And I think the documentation needs lots of improvements, most of the API and use cases are not covered yet.

**Comment by reintroducing** [comment]
you can use the (for some reason undocumented) `useLocation` hook to access the current route.

**Comment by Sawtaytoes** [comment]
That hook no longer exists. That's probably why it wasn't documented :P.

I'm also wanting an answer to this question. How do you grab the current route?

**Comment by 0x2f0** [comment]
I am also searching for a way to grab the currently active route,
I can use the window location api to get the entire path and then get the active path but having a default trackable way would be great.

**Comment by ParasSolanki** [comment]
For the current active route state you have to use `useRouterState` (not documented yet).

https://github.com/TanStack/router/blob/a542861cf430d5a7f0827904dbbb424340b957d5/packages/react-router/src/useRouterState.tsx#L6

```tsx
// ./src/routes/activity.lazy.tsx
import { createLazyFileRoute, useRouterState } from "@tanstack/react-router";

export const Route = createLazyFileRoute("/activity")({
  component: ActivityPage,
});


function ActivityPage() {
  const state = useRouterState();
}
```

**Reply by SeanCassiere** [reply, marked as answer]
I'd recommend using a more fine-grained selector to reduce the number of component re-renders.

```tsx
const location = useRouterState({ select: (s) => s.location })
```

**Reply by ParasSolanki** [reply]
Yeah, selector api is great.
